File: Object_Oriented_Design/Card.java
import java.util.HashSet;

public class Card {

    // Create class variables
    String suit;
    String name;
    String color;
    int value;

    void setSuit(String suit) {

        // OBJECTIVE: Set card's suit based on "suit" variable

        // List suit types
        HashSet<String> suits = new HashSet<>();
        suits.add("Clubs");
        suits.add("Hearts");
        suits.add("Spades");
        suits.add("Diamonds");

        // If suit is in suits, update suit
        if (suits.contains(suit)) {this.suit = suit;}
    }

    void setName(String name) {

        // OBJECTIVE: Set card name based on "name" variable

        // List name options
        HashSet<String> names = new HashSet<>();
        names.add("Jack");
        names.add("Queen");
        names.add("King");
        names.add("Ace");

        // If name is part of set, update name variable
        if (names.contains(name)) {
            this.name = name;
        }
        
        // Check if name is numeric
        try {
            this.name = Integer.toString(Integer.parseInt(name));
        }
        catch (NumberFormatException err){
            ; // <= Similar to Python's "pass"
        }

        // If function is still going, then set card as invalid
        this.name = "INVALID CARD!";
    }

    void setColor() {

        // OBJECTIVE: Set card's color based on suit type

        if (this.suit == "Clubs" || this.suit == "Spades") {
            this.color = "Black";
        }
        else if (this.suit == "Hearts" || this.suit == "Diamonds") {
            this.color = "Red";
        }
    }

    void setValue(int value) {

        // OBJECTIVE: Set card's color based on card's name

        if (this.name == "Jack") {
            this.value = 11;
        }

        else if (this.name == "Queen") {
            this.value = 12;
        }

        else if (this.name == "King") {
            this.value = 13;
        }

        // If card's name is Ace, set value to either 1 or 11
        else if (this.name == "Ace") {
            this.value = value == 1 || value == 11 ? value : 1; 
        }

        else {

            try {

                // Convert string to an integer
                int tmpValue = Integer.valueOf(value);

                // Keep input if it's between 1-10
                if (tmpValue > 1 || tmpValue < 10) {
                    this.value = tmpValue;
                }
                else {
                    this.value = -1;
                }
            }
            catch (NumberFormatException err) {
                ;
            }
        }
    }

    Card(String suit, String name) {

        // OBJECTIVE: Construct class without given input value. If so, then this must be a face card

        setSuit(suit);
        setName(name);
        setColor();
        setValue(0);
    }

    Card(String suit, String name, int value) {

        // OBJECTIVE: Construct class with given input value. If so, this isn't a face card

        setSuit(suit);
        setName(name);
        setColor();
        setValue(value);
    }
}

File: Object_Oriented_Design/Card.py
# Cracking the Coding Interview - pp. 127 - q 7.1

class Card:

    def __init__(self, name, color, value) -> None:
        self.name = name
        self.color = color
        self.value = value

class AdvancedCard:

    def setSuit(self, suit):

        # OBJECTIVE: Set card's suit based on "suit" variable

        if suit in {"Clubs", "Hearts", "Spades", "Diamonds"}:
            self.suit = suit

    def setName(self, name):

        # OBJECTIVE: Set card name based on "name" variable

        # If name is in set, set name variable
        if name in {"Jack", "Queen", "King", "Ace"}:
            self.name = name

        # If name is numeric only, set name variable
        elif name.isnumeric():
            self.name = name

        # If all above conditions are false, set card as invalid
        else:
            self.name = "INVALID CARD!"

    def setColor(self):

        # OBJECTIVE: Set card's color based on suit type

        if self.suit == "Clubs" or self.suit == "Spades":
            self.color = "Black"

        elif self.suit == "Hearts" or self.suit == "Diamonds":
            self.color = "Red"

    def setValue(self, value):

        # OBJECTIVE: Set card's color based on card's name

        if self.name == "Jack":
            self.value = 11
        
        elif self.name == "Queen":
            self.value = 12
        
        elif self.name == "King":
            self.value = 13

        elif self.name == "Ace":

            # If card's name is Ace and input "value" is either 1 or 11, set it as input.
            # If an invalid value was given (e.g., 13), set card's value to 1 as default
            self.value = value if value == 1 or value == 11 else 1

        # Change "name" to an integer. NOTE: In setName(), if "name" is numeric, integer conversion won't throw an error
        elif self.name.isnumeric():
            tmpValue = int(self.name)

            if tmpValue > 1 and tmpValue < 10:
                self.value = tmpValue
            else:
                self.value = -1

    def __init__(self, suit, name, value=0):

        self.setSuit(suit)
        self.setName(name)
        self.setColor()
        self.setValue(value)

    def info(self):
        print("{} {} {} {}".format(self.color, self.suit, self.name, self.value))

def main():

    one = AdvancedCard("Clubs", "King")
    two = AdvancedCard("Hearts", "Jack")
    three = AdvancedCard("Spades", "Queen")
    four = AdvancedCard("Diamonds", "Ace")
    five = AdvancedCard("Hearts", "1")

    one.info()
    two.info()
    three.info()
    four.info()
    five.info()

main()